### PR TITLE
Fix subcontext metric space

### LINKED ISSUE #1375: Context API: track expected metric space for queries separately from current metric space
```python
import opendp.prelude as dp
dp.enable_features("contrib")

# input metric space is (vector domain, symmetric distance)
context = dp.Context.compositor(
    data=[1] * 100,
    privacy_unit=dp.unit_of(contributions=1), # This makes it symmetric distance
    privacy_loss=dp.loss_of(epsilon=1.),
    split_evenly_over=1
)

# A case where this fix would be useful:
#   The top-level privacy is approx-dp, but the lower-level compositor is zc-DP.

# input metric space on the compositor is still (vector domain, symmetric distance),
# but the queries you pass into the compositor should be on (atom domain, absolute distance)
subcontext = context.query().clamp((0, 1)).sum().compositor(split_evenly_over=1).release()

# The above runs, but has bad internal state in the compositor.

# therefore down here, the chain should initialize to (atom domain, absolute distance)
# instead of (vector domain, symmetric distance)
# Input space for queries is always the input space of compositor: Not what we want!
# Compositors may expect queries in a different space, as in the example above.
print(subcontext.query()._chain)

# Right now, you'll see the (vector, symmetric)
# but we want the output space of the operation preceding the compositor.

# in most settings, the input space of the current compositor 
#     has happened to match the expected input space of queries
# to fix, could also store the expected input space of queries in the context

# this also fails! see issue #1377
subcontext.query().laplace()
```

### LINKED ISSUE #1376: Context API: Return type hint on Query.compositor should be "Query"

### PR DESCRIPTION
**Pull request by mccalluc**
- Fix #1375
- Fix #1376

Just some reprs right now, but wanted to share this much. Draft.

### CODE REVIEW

**Review by Shoeboxam — COMMENTED** [review]
Partial review, just sharing thoughts before our meeting.

File: python/test/context/test_context.py [hunk 1]
```diff
@@ -5,9 +5,24 @@
...
+    assert dp.unit_of(contributions=3, ordered=True) == (dp.insert_delete_distance(), 3)
+
+    assert dp.unit_of(changes=3) == (dp.change_one_distance(), 3)
+    assert dp.unit_of(changes=3, ordered=True) == (dp.hamming_distance(), 3)
+
+    # TODO: For the cases below, is ordered meaningful?
+    # Should an exception be raised?
+
+    assert dp.unit_of(absolute=3) == (dp.absolute_distance(dp.i32), 3)
+    assert dp.unit_of(absolute=3, ordered=True) == (dp.absolute_distance(dp.i32), 3)
```
Comment (mccalluc) [inline python/test/context/test_context.py:19]: Make this a value error.

File: python/test/context/test_context.py [hunk 2]
```diff
@@ -52,6 +87,21 @@ def test_context_init_split_evenly_over():
     # print(context.query().dp_sum((1, 10)).release())
 
 
+def test_partial_chain_right_shift():
```
Comment (mccalluc) [inline python/test/context/test_context.py:90]: Just drop for now.

File: python/test/context/test_context.py [hunk 3]
```diff
@@ -67,32 +117,89 @@ def test_context_zCDP():
...
+
+
 def test_sc_query():
     context = dp.Context.compositor(
         data=[1, 2, 3],
         privacy_unit=dp.unit_of(contributions=1),
         privacy_loss=dp.loss_of(epsilon=3.0, delta=1e-6),
         split_evenly_over=2,
+        # TODO: Using split_by_weights instead fails:
+        # split_by_weights=[1, 2],
```
Comment (mccalluc) [inline python/test/context/test_context.py:182]: It should work! file issue / make a bug test.

File: python/src/opendp/context.py [hunk 4]
```diff
@@ -521,16 +543,19 @@ def release(self) -> Any:
...
         split_evenly_over: Optional[int] = None,
         split_by_weights: Optional[List[float]] = None,
         d_out=None,
         output_measure=None,
-    ) -> "Context":
+    ) -> "Query":
         """Constructs a new context containing a sequential compositor with the given weights.
+        
+        .. TODO: I think it actually returns a Query,
+        .. or at least changing the type fixes mypy, and agrees with the repr.
```
[the remark this answers is not in the document]
Comment (mccalluc) [reply python/src/opendp/context.py:558]: Thanks -- Add a reference at the top of this PR

File: python/test/context/test_context.py [hunk 5]
```diff
@@ -24,6 +39,26 @@ def test_option_domain():
...
+            privacy_loss=dp.loss_of(epsilon=3.0),
+            split_evenly_over=1,
+            domain=dp.domain_of(List[int]),
+        )
+    ) == '''Context(
+    accountant = Measurement(
+        input_domain   = VectorDomain(AtomDomain(T=i32)),
+        input_metric   = SymmetricDistance(),
+        output_measure = MaxDivergence(f64)),
+    queryable  = Queryable(Q=AnyMeasurement),
```
[the remark this answers is not in the document]
Comment (mccalluc) [reply python/test/context/test_context.py:56]: If I'm understanding your preference, the request is to drop just queryable, so the assertion would be:
```
    ) == '''Context(
    accountant = Measurement(
        input_domain   = VectorDomain(AtomDomain(T=i32)),
        input_metric   = SymmetricDistance(),
        output_measure = MaxDivergence(f64)),
    d_in       = 3,
    d_mids     = [3.0],
    d_out      = None)'''
```

File: python/test/context/test_context.py [hunk 6]
```diff
@@ -24,6 +39,26 @@ def test_option_domain():
...
+    d_in       = 3,
+    d_mids     = [3.0],
+    d_out      = None)'''
```
[the remark this answers is not in the document]
Comment (mccalluc) [reply python/test/context/test_context.py:59]: so maybe I'll drop `d_out` from the repr for now, and it can be added back when it might have useful info?

**Review by Shoeboxam — APPROVED** [review]
Thanks for digging to debug this!

**Pull request merged**